pilot/pkg/xds/delta.go

```diff
@@ -0,0 +1,590 @@
+	con.proxy.WatchedResources[request.TypeUrl].NonceAcked = request.ResponseNonce
+	con.proxy.WatchedResources[request.TypeUrl].NonceNacked = ""
+	con.proxy.WatchedResources[request.TypeUrl].ResourceNames = deltaWatchedResources(previousResources, request)
+	con.proxy.WatchedResources[request.TypeUrl].LastRequest = deltaToSotwRequest(request)
+	con.proxy.Unlock()
+
+	oldAck := listEqualUnordered(previousResources, con.proxy.WatchedResources[request.TypeUrl].ResourceNames)
+	newAck := request.ResponseNonce != ""
+	if newAck != oldAck {
+		// Not sure which is better, lets just log if they don't match for now and compare.
+		adsLog.Errorf("dADS:%s: New ACK and old ACK check mismatch: %v vs %v", stype, oldAck, newAck)
+	}
+	// Envoy can send two DiscoveryRequests with same version and nonce
+	// when it detects a new resource. We should respond if they change.
+	if oldAck {
+		adsLog.Debugf("dADS:%s: ACK %s %s", stype, con.ConID, request.ResponseNonce)
+		return false
+	}
+	adsLog.Debugf("dADS:%s: RESOURCE CHANGE previous resources: %v, new resources: %v %s %s", stype,
+		previousResources, con.proxy.WatchedResources[request.TypeUrl].ResourceNames, con.ConID, request.ResponseNonce)
+
+	return true
+}
+
+// Push an XDS resource for the given connection. Configuration will be generated
+// based on the passed in generator. Based on the updates field, generators may
+// choose to send partial or even no response if there are no changes.
+func (s *DiscoveryServer) pushDeltaXds(con *Connection, push *model.PushContext, currentVersion string,
+	w *model.WatchedResource, subscribe []string, req *model.PushRequest) error {
+	if w == nil {
+		return nil
+	}
+	gen := s.findGenerator(w.TypeUrl, con)
+	if gen == nil {
+		return nil
+	}
+
+	t0 := time.Now()
+
+	res, err := gen.Generate(con.proxy, push, w, req)
+	if err != nil || res == nil {
+		// If we have nothing to send, report that we got an ACK for this version.
+		if s.StatusReporter != nil {
+			s.StatusReporter.RegisterEvent(con.ConID, w.TypeUrl, push.LedgerVersion)
+		}
+		return err
+	}
+	defer func() { recordPushTime(w.TypeUrl, time.Since(t0)) }()
+
+	deltaResponse := convertResponseToDelta(currentVersion, res)
+	originalResponse := deltaResponse
+	if subscribe != nil {
+		subres := sets.NewSet(subscribe...)
+		filteredResponse := []*discovery.Resource{}
+		for _, r := range deltaResponse {
+			if subres.Contains(r.Name) {
+				filteredResponse = append(filteredResponse, r)
+			} else {
+				adsLog.Debugf("ADS:%v SKIP %v", v3.GetShortType(w.TypeUrl), r.Name)
+			}
+		}
+		deltaResponse = filteredResponse
+	}
+	resp := &discovery.DeltaDiscoveryResponse{
+		TypeUrl:           w.TypeUrl,
+		SystemVersionInfo: currentVersion,
+		Nonce:             nonce(push.LedgerVersion),
+		Resources:         deltaResponse,
+	}
+	cur := sets.NewSet(w.ResourceNames...)
+	cur.Delete(extractNames(originalResponse)...)
+	resp.RemovedResources = cur.SortedList()
```

Comment:
not sure i understand this removed resource cal logic

Reply:
+1